iina/JavasctiptDevTool.swift

```diff
@@ -0,0 +1,506 @@
+//
+//  Created by Hechen Li on 11/30/22.
+//  Copyright © 2022 lhc. All rights reserved.
+//
+
+import SwiftUI
+import JavaScriptCore
+import WebKit
+
+
+// MARK: - Menu Controller
+
+@available(macOS 12.0, *)
+extension MenuController {
+  func createMenuItem(fromPluginInstance inst: JavascriptPluginInstance, tag: Int) -> NSMenuItem {
+    let item = NSMenuItem()
+    if inst.isGlobal {
+      item.title = "\(inst.plugin.name) (Global)"
+    } else {
+      item.title = inst.plugin.name
+    }
+    item.representedObject = inst
+    item.tag = JavasctiptDevTool.JSMenuItemInstance
+    item.target = self
+    item.action = #selector(openJavascriptDevTool)
+    return item
+  }
+
+  @objc func openJavascriptDevTool(_ sender: NSMenuItem) {
+    switch (sender.tag) {
+    case JavasctiptDevTool.JSMenuItemInstance:
+      if let inst = sender.representedObject as? JavascriptPluginInstance {
+        createJavascriptDevToolWindow(forInstance: inst, title: sender.title)
+      }
+      //    case JSMenuItemWebView:
+      //      is let webView = sender.representedObject as? WKWebView {
+      //        createJavascriptDevToolWindow(forWebView: webView)
+      //      }
+    default:
+      break
+    }
+  }
+}
+
+
+// MARK: - Data
+
+fileprivate struct JSEvent: Identifiable {
+  enum Result {
+    case nothing
+    case number(NSNumber)
+    case string(String)
+    case boolean(Bool)
+    case array(String, [(String, String)])
+    case null
+    case undefined
+    case opaqueObject(String)
+    case object(String, [(String, String)])
+    case exception(String, String)
+    case log(String, Logger.Level)
+  }
+  let id: Int
+  let index: Int
+  let prompt: String?
+  let result: Result
+
+  var isMessage: Bool {
+    switch self.result {
+    case .exception(_, _), .log(_, _):
+      return true
+    default:
+      return false
+    }
+  }
+}
+
+@available(macOS 12.0, *)
+fileprivate class JSEventsContainer : ObservableObject{
+  @Published var idCounter = 0
+  @Published var counter = 0
+  @Published var data = [JSEvent]()
+  @Published var idChanged = 0
+
+  func addPrompt(_ prompt: String) {
+    idCounter += 1
+    counter += 1
+    data.append(JSEvent(id: idCounter, index: counter, prompt: prompt, result: .nothing))
+  }
+
+  func addResult(_ result: JSEvent.Result) {
+    idCounter += 1
+    data.append(JSEvent(id: idCounter, index: counter, prompt: nil, result: result))
+    postChange()
+  }
+
+  func addException(message: String, stack: String) {
+    idCounter += 1
+    data.append(
+      JSEvent(id: idCounter, index: counter, prompt: nil,
+                result: .exception(message, stack)))
+    postChange()
+  }
+
+  func addLog(message: String, level: Logger.Level) {
+    idCounter += 1
+    data.append(
+      JSEvent(id: idCounter, index: counter, prompt: nil,
+                result: .log(message, level)))
+    postChange()
+  }
+
+  func postChange() {
+    Timer.scheduledTimer(withTimeInterval: 0.1, repeats: false) { _ in
```

Comment:
Do I want to know

Reply:
This is used to scroll the ScrollView to the bottom. For some reason (I can't remember exactly) it doesn't work when called immediately.